launch_testing/examples/test_pub_sub_launch.py

```diff
@@ -0,0 +1,230 @@
+    # launch service in some way to ease launch file composition without
+    # repetition).
+    ls.include_launch_description(
+        get_default_launch_description(prefix_output_with_name=False)
+    )
+    return ls
+
+
+@pytest.fixture
+def pub_sub_launch_description() -> LaunchDescription:
+    """
+    A test fixture providing the basic pub/sub launch description.
+    """
+    # TODO(hidmic): implement functionality to lookup launch files
+    # on a per package basis
+    launch_path = get_launch_path(
+        package_name='launch_ros', launch_file_path='pub_sub_launch.py'
+    )
+    return get_launch_description_from_python_launch_file(launch_path)
+
+
+def test_pub_sub_launch(ros_test_launch_service, pub_sub_launch_description):
+    """
+    Tests an existing simple pub/sub system.
+    """
+    # Launches a pytest that also happens to be a node with
+    # a 30s timeout.
+    pub_sub_launch_description.add_action(
+        # TODO(hidmic): allow launching non-installed
+        # executables within the current package.
+        PyTest(Node(
+            node_executable='pub_test.py',
+            remappings=[('chatter', 'my_chatter')]
+        ), timeout=30)
+    )
+    # Launches a gtest that also happens to be a node with
+    # a 30s timeout.
+    pub_sub_launch_description.add_action(
+        GTest(Node(
+            node_executable='sub_test',
+            remappings=[('chatter', 'my_chatter')]
+        ), timeout=30)
+    )
+
+    ros_test_launch_service.include_launch_description(pub_sub_launch_description)
+
+    # TODO(hidmic): implement launch_testing specific pytest plugin to aggregate
+    # all test result information at the launch system-level and below.
+    assert ros_test_launch_service.run() == 0
+
+
+@pytest.fixture
+def lifecycle_pub_sub_launch_description() -> LaunchDescription:
+    """
+    A test fixture providing the lifecycle pub/sub launch description.
+    """
+    # TODO(hidmic): implement functionality to lookup launch files
+    # on a per package basis
+    launch_path = get_launch_path(
+        package_name='launch_ros',
+        launch_file_path='lifecycle_pub_sub_launch.py'
+    )
+    return get_launch_description_from_python_launch_file(launch_path)
+
+
+def test_lifecycle_pub_sub_launch(ros_test_launch_service,
+                                  lifecycle_pub_sub_launch_description):
+    """
+    Tests an existing lifecycle pub/sub nodes setup for execution
+    in proper sequence.
+    """
+    # Asserts that a lifecycle.ChangeState event within a 30s window
+    # is followed by a process.ProcessStarted event within a 10s window
+    # once (40s total timeout).
+    lifecycle_pub_sub_launch_description.add_action(
+        AssertSequenceOnce([
+            variables.EventEmitted(
+                event_type=lifecycle.ChangeState
+            ),
+            variables.EventEmitted(
+                event_type=ProcessStarted
+            )
+        ], timeout=[30, 10])
```

Comment:
> This is interesting, and I like it, but the the generic name really needs a comment

Sure. I'm open to any naming changes. How would you briefly state that you're asserting that N conditions occur in the right sequence/order?

> Also, would it be valuable to ensure that the events came from a specific process?

For sure. Don't take these examples as a completely exhaustive demonstration of all the (future) API features. In this case, we could as easily add some matching to only check for the source (i.e. the executable) of interest.

> In which case, would it make sense to couple this assertion to the thing being tested?

I did play with this idea. It wasn't clear to me either how to deal with multiple assertions of different kinds, potentially targeting multiple actions. That's why I dropped it in favor of using the target action instances directly in the predicates being asserted. We could have something similar to what you describe though, by deferring variable binding. E.g. instead of:

```python
Assert(len(predicates.regex_match(
    variables.Output(target), pattern='^Oops!.*'
)) > 10, timeout=10)
```

we could write:

```python
Test(target=Node(), description=[
    Assert(len(predicate.regex_match(
        variables.Output(), pattern='Oops!.*'
    )) > 10, timeout=10)
])
```

It may get somewhat confusing though if you start mixing both on the same launch file. The former is the most general one, so I don't think we can replace it with the latter.

Reply:
> Sure. I'm open to any naming changes. How would you briefly state that you're asserting that N conditions occur in the right sequence/order?

I don't have an alternative in mind, but I'll keep thinking on it.

The rest makes sense to me, you can realize those cases when/if we need them.